src/derivative_operators/convolutions.jl

```diff
@@ -43,34 +65,65 @@ end
 function convolve_BC_right!(x_temp::AbstractVector{T}, x::AbstractVector{T}, A::DerivativeOperator) where {T<:Real}
     stencil = A.high_boundary_coefs
     coeff   = A.coefficients
-    for i in 1 : A.boundary_point_count
-        cur_stencil = stencil[i]
+    N       = length(x)
+    L       = A.boundary_stencil_length
+
+    _bpc = A.boundary_point_count
+    flag = false
+
+    if stencil == []
+        _bpc = 1
+        flag = true
+    end
+
+    for i in 1 : _bpc
+        # @show _bpc
+        if flag == true
+            cur_stencil = A.stencil_coefs
+            slen = length(A.stencil_coefs)
+            L = A.stencil_length
+        else
+            cur_stencil = stencil[i]
+            slen = length(cur_stencil)
+        end
+
         cur_coeff   = typeof(coeff)   <: AbstractVector ? coeff[i] : coeff isa Number ? coeff : true
         cur_stencil = use_winding(A) && cur_coeff < 0 ? reverse(cur_stencil) : cur_stencil
-        xtempi = cur_coeff*stencil[i][end]*x[end]
-        for idx in (A.boundary_stencil_length-1):-1:1
-            xtempi += cur_coeff * cur_stencil[end-idx] * x[end-idx]
+        cs = cur_stencil*(A.dx^A.derivative_order)
+        xtempi = zero(T)
+        for idx in 1:slen
+            # @show idx, N-L+idx, cs[idx], x[N-L+idx]
+            xtempi += cur_coeff * cur_stencil[idx] * x[N-L+idx]
         end
-        x_temp[end-A.boundary_point_count+i] = xtempi
+        # @show xtempi*(1/(A.dx^A.derivative_order))
+        x_temp[end-_bpc+i] = xtempi
     end
 end
 
 ###########################################
 
 # Against A BC-padded vector, specialize the computation to explicitly use the left, right, and middle parts
 function convolve_interior!(x_temp::AbstractVector{T}, _x::BoundaryPaddedVector, A::DerivativeOperator) where {T<:Real}
+    @assert length(x_temp) == length(_x.u)
     stencil = A.stencil_coefs
     coeff   = A.coefficients
-    x = _x.u
-    mid = div(A.stencil_length,2) + 1
+    # Upwind operators have a non-centred stencil
+    if use_winding(A)
+        mid = 1 + A.stencil_length%2
+    else
+        mid = div(A.stencil_length,2)
+    end
+
     # Just do the middle parts
-    for i in (2+A.boundary_point_count) : (length(x_temp)-A.boundary_point_count)-1
+    for i in (1+A.boundary_point_count) : (length(x_temp)-A.boundary_point_count)
         xtempi = zero(T)
         cur_stencil = eltype(stencil) <: AbstractVector ? stencil[i-A.boundary_point_count] : stencil
         cur_coeff   = typeof(coeff)   <: AbstractVector ? coeff[i-A.boundary_point_count] : coeff isa Number ? coeff : true
         cur_stencil = use_winding(A) && cur_coeff < 0 ? reverse(cur_stencil) : cur_stencil
         @inbounds for idx in 1:A.stencil_length
-            xtempi += cur_coeff * cur_stencil[idx] * x[(i-1) - (mid-idx) + 1]
+            # @show i, idx, cur_stencil[idx], i-mid+idx, x[i-mid+idx]
+            x_idx = use_winding(A) && cur_coeff < 0 ? _x[i + mid - idx] : _x[i - mid + idx]
```

Comment:
@ChrisRackauckas @shivin9 Here we are indexing into `_x`, a BoundaryPaddedVector using the slower fallback indexing. Should we instead be indexing into `_x.u` as was implemented earlier?

Reply:
Yeah, ideally that should be done...

Reply:
This is the code benchmark for the two implementations, is it worth sacrificing code readability?
```
# using _x.u, _x.l etc.
julia> @benchmark fourth_deriv_approx_stencil(N)*arrayQu ≈ L*Qu
BenchmarkTools.Trial: 
  memory estimate:  8.03 KiB
  allocs estimate:  32
  --------------
  minimum time:     2.874 μs (0.00% GC)
  median time:      3.118 μs (0.00% GC)
  mean time:        5.012 μs (31.69% GC)
  maximum time:     7.953 ms (99.85% GC)
  --------------
  samples:          10000
  evals/sample:     9

# Current way
julia> @benchmark fourth_deriv_approx_stencil(N)*arrayQu ≈ L*Qu
BenchmarkTools.Trial: 
  memory estimate:  8.03 KiB
  allocs estimate:  32
  --------------
  minimum time:     3.089 μs (0.00% GC)
  median time:      3.301 μs (0.00% GC)
  mean time:        5.257 μs (32.65% GC)
  maximum time:     8.887 ms (99.87% GC)
  --------------
  samples:          10000
  evals/sample:     8
```